src/wallet/rpcdump.cpp

```diff
@@ -651,6 +651,59 @@ UniValue z_importkey(const UniValue& params, bool fHelp)
     return NullUniValue;
 }
 
+UniValue z_validateviewingkey(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    if (fHelp || params.size() != 1)
+        throw runtime_error(
+            "z_validateviewingkey \"vkey\"\n"
+            "\nReturn information about the given Zcash viewing key.\n"
+            "\nArguments:\n"
+            "1. \"vkey\"             (string, required) The viewing key (see z_exportviewingkey)\n"
+            "\nExamples:\n"
+            "\nValidate a viewing key\n"
+            + HelpExampleCli("z_validateviewingkey", "\"vkey\"") +
+            "\nAs a JSON-RPC call\n"
+            + HelpExampleRpc("z_validateviewingkey", "\"vkey\"")
+        );
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    EnsureWalletIsUnlocked();
+
+    UniValue ret(UniValue::VOBJ);
+    string strVKey  = params[0].get_str();
+    auto viewingkey = DecodeViewingKey(strVKey);
+
+    if (!IsValidViewingKey(viewingkey)) {
+	ret.push_back(Pair("isvalid", false));
+	return ret;
+    }
+
+    if( boost::get<libzcash::SproutViewingKey>(&viewingkey) != nullptr ) {
```

Comment:
Use a `boost::static_visitor` to ensure all cases are handled (both now and later). See #3345 for an example. The code in this function will end up looking something like:

```cpp
    auto vkey = DecodeViewingKey(viewingkey);
    bool isValid = IsValidPaymentAddress(vkey);

    UniValue ret(UniValue::VOBJ);
    ret.push_back(Pair("isvalid", isValid));
    if (isValid) {
        UniValue detail = boost::apply_visitor(DescribeViewingKeyVisitor(), vkey);
        ret.pushKVs(detail);
    }
    return ret;
```

Reply:
@str4d thanks for the feedback, I will work on making this RPC Sapling-aware